Author: Jamie Strandboge <jamie@canonical.com>
Description: Adjust apparmor(5) to describe policy load on Ubuntu
Bug-Ubuntu: https://launchpad.net/bugs/974089
Forward: no (Ubuntu specific)

---
 parser/apparmor.pod |   44 +++++++++++++++++++++++++++++++++++++++++---
 1 file changed, 41 insertions(+), 3 deletions(-)

Index: apparmor-2.8.96~2652/parser/apparmor.pod
===================================================================
--- apparmor-2.8.96~2652.orig/parser/apparmor.pod
+++ apparmor-2.8.96~2652/parser/apparmor.pod
@@ -3,7 +3,7 @@
 #                  2008, 2009
 #    NOVELL (All rights reserved)
 #
-#    Copyright (c) 2010
+#    Copyright (c) 2010, 2011, 2012
 #    Canonical Ltd. (All rights reserved)
 #
 #    Copyright (c) 2013
@@ -19,7 +19,7 @@
 #    GNU General Public License for more details.
 #
 #    You should have received a copy of the GNU General Public License
-#    along with this program; if not, contact Novell, Inc.
+#    along with this program; if not, contact Canonical Ltd.
 # ----------------------------------------------------------------------
 
 
@@ -37,7 +37,8 @@ attributes to programs rather than to us
 
 AppArmor confinement is provided via I<profiles> loaded into the kernel
 via apparmor_parser(8), typically through the F</etc/init.d/apparmor>
-SysV initscript, which is used like this:
+SysV initscript (on Ubuntu, see UBUNTU POLICY LOAD, below), which is used like
+this:
 
 	# /etc/init.d/apparmor start
 	# /etc/init.d/apparmor stop
@@ -95,6 +96,62 @@ cannot call the following system calls:
 	iopl(2) ptrace(2) reboot(2) setdomainname(2)
 	sethostname(2) swapoff(2) swapon(2) sysctl(2)
 
+=head1 UBUNTU POLICY LOAD
+
+Ubuntu systems use upstart(8) instead of a traditional SysV init system.
+Because upstart is an event-driven init system and understanding that policy
+must be loaded before execution, Ubuntu loads policy in two ways:
+
+ 1. via upstart jobs for services started during the boot process
+ 2. via the AppArmor upstart job for any remaining policy
+
+The AppArmor upstart job is configured to make sure all policy is loaded before
+any user sessions start. When developing policy it is important to know how
+your application is started and if policy load should be handled specially.
+
+The upstart job may be call with or without arguments, like so:
+
+	$ sudo start apparmor
+	$ sudo start apparmor ACTION=clear         # clear policy cache
+	$ sudo start apparmor ACTION=teardown      # unload all policy
+	$ sudo start apparmor ACTION=reload        # reload policy
+	$ sudo start apparmor ACTION=force-reload  # same as 'reload'
+
+Because the job is an upstart(8) task, use 'start apparmor ACTION=teardown' to
+unload all policy.
+
+In general, nothing extra has to be done for applications not started during
+boot or those that start after AppArmor's upstart job.
+
+If the confined application has an Upstart job, adjust the job to call
+F</lib/init/apparmor-profile-load> with the filename of the policy file
+(relative to F</etc/apparmor.d/>). For example:
+
+	pre-start script
+    		/lib/init/apparmor-profile-load usr.bin.foo
+	end script
+
+If the confined application does not have an Upstart job but it starts before
+AppArmor's second stage initscript, then add a symlink from the policy file in
+F</etc/apparmor.d> to F</etc/apparmor/init/network-interface-security/>. For
+example:
+
+	$ cd /etc/apparmor/init/network-interface-security/
+	$ sudo ln -s /etc/apparmor.d/usr.bin.foo .
+
+The network-interface-security Upstart job will load all the symlinked policy
+files in F</etc/apparmor/init/network-interface-security/> before any network
+interfaces come up. Because network interfaces come up very early in the boot
+process, this will help ensure that AppArmor policy is loaded before the
+confined application starts.
+
+In addition, AppArmor on Ubuntu stores policy in two places:
+
+ 1. /etc/apparmor.d for system policy
+ 2. /var/lib/apparmor/profiles for click policy
+
+See apparmor_parser(8) and aa-clickhook(1) for details.
+
 =head1 ERRORS
 
 When a confined process tries to access a file it does not have permission
@@ -146,6 +203,8 @@ depending upon local configuration.
 
 =item F</etc/init.d/apparmor>
 
+=item F</etc/apparmor/init/network-interface-security/>
+
 =item F</etc/apparmor.d/>
 
 =item F</var/lib/apparmor/>
